proposals/0000-mutable-fields.rst

```diff
@@ -0,0 +1,249 @@
+back to that later)
+
+But what *are* these ``Ref#`` things that are extracted from the
+constructor and seemingly passed to ``readRef#``?  The idea is that the
+runtime representation of ``Ref#`` is a pair of the containing object
+and the offset of the mutable field. Just before code generation,
+probably in the Unarise phase, we will manifest the runtime
+representation of ``Ref#`` at each pattern match::
+
+  readMutPair m = IO $ \s0 ->
+    case m of (v::Any) { MutPair aref bref ->
+    let aref = (# v, 0# #) in
+    let bref = (# v, 1# #) in
+    case readRef# aref s0 of { (# s1, a #) ->
+    case readRef# bref s1 of { (# s2, b #) -> (# s2, (a, b) #) }}}
+
+and then propagate the expansion of ``aref`` and ``bref`` to all the
+places they are referenced. Function arguments of type ``Ref#`` are
+expanded to unboxed pairs of type ``(# Any, Int# #)``.
+
+The ``readRef#`` primitive would be compiled inline to a single read
+instruction. Similarly ``writeRef#`` would compile to a primitive write
+instruction, but it would also need a memory barrier just like
+``writeMutVar#``, and a GC write barrier (the equivalent of
+``dirty_MUT_VAR()``).
+
+Simpifying the programmer's API
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+To make it easier for the programmer we would wrap the readRef#
+primitive::
+
+  readRef :: IOField a -> IO a
+  readRef aref = IO $ \s -> readRef# aref
+
+And then we could write::
+
+  readMutPair :: MutPair a b -> IO (a, b)
+  readMutPair (MutPair aref bref) = (,) <$> readRef aref <*> readRef bref
+
+Garbage collection
+~~~~~~~~~~~~~~~~~~
+
+The garbage collector needs to know that an object is mutable, and
+which fields are mutable.  So we have to put this information in the
+info table.  Probably:
+
+- new ``MUT_CONSTR_CLEAN`` and ``MUT_CONSTR_DIRTY`` object types
+
+- Each constructor with a mutable field needs two info tables: the
+  CLEAN one and the DIRTY one.  From each info table we need to be
+  able to get both CLEAN and DIRTY info pointers, for the GC write
+  barrier.
+
+- Representation: put the mutable fields first, before the non-mutable
+  pointers.
+
+- Store the number of mutable fields in the info table, alongside the
+  number of pointers and non-pointers.  (but include the mutable
+  fields in the count of pointer fields, so that RTS code that doesn't
+  care about mutability can work without changes)
+
+The GC needs to do the same CLEAN/DIRTY and non-eager promotion stuff
+that it does with other mutable objects.
+
+Code generation
+~~~~~~~~~~~~~~~
+
+We would generate code for the primitive constructor just like we
+generate code for other constructors, taking care to add the Void
+argument for the ``State#``, and generating an info table with the
+correct information about the mutable fields.
+
+Unpacking constructors with mutable fields
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+UNPACK must be a no-op on constructors with mutable fields.  There's
+no sensible way to make UNPACK work with mutable fields.
```

Comment:
I believe this refers to unpacking a mutable constructor into something else.  But it would be nice to unpack the reasoning there.  I think it goes like this:

"While C++ or Rust allow unpacking a mutable product type into another, this requires a recursive notion of object construction/initialization and would not fit here.  If we tried the following:

```Haskell
data MutPair2 = MP Int {-# UNPACK #-} MutPair
```

what is the type of the `MP` constructor?  It cannot be this:

```Haskell
MP :: Int -> MutPair -> IO MutPair2
```

because the only way to make that work would be to *copy* the mutable record `MutPair` into the `MP` constructor.  This is (1) not what we'd want for including inner objects as value types, and (2) ruins the guarantee that `UNPACK` is a performance hint rather than semantically important."